src/spreadinterp.cpp

```diff
@@ -918,237 +982,458 @@ void interp_cube(FLT *target, FLT *du, FLT *ker1, FLT *ker2, FLT *ker3, BIGINT i
+  static constexpr auto alignment = batch_t::arch_type::alignment();
+  static constexpr auto avx_size  = batch_t::size;
+  static constexpr auto ns2       = ns * FLT(0.5); // half spread width
+  // something weird here. Reversing ker{0} and std fill causes ker
+  // to be zeroed inside the loop GCC uses AVX, clang AVX2
+  std::fill(du, du + 2 * size1, 0); // zero output
+  // no padding needed if MAX_NSPREAD is 16
+  // the largest read is 16 floats with avx512
+  // if larger instructions will be available or half precision is used, this should be
+  // padded
+  for (uint64_t i{0}; i < M; i++) { // loop over NU pts
+    // lamda here to return a dd_pt that is const
+    // should not make a difference in performance
+    // but is a hint to the compiler that after the lambda
+    // dd_pt is not modified and can be kept as is in a register
+    // given (re, im) in this case dd[i*2] and dd[i*2+1]
+    // this function returns a simd register of size avx_size
+    // initialized as follows:
+    // +-----------------------+
+    // |re|im|re|im|re|im|re|im|
+    // +-----------------------+
+    const auto dd_pt = initialize_complex_batch<batch_t>(dd[i * 2], dd[i * 2 + 1]);
     // ceil offset, hence rounding, must match that in get_subgrid...
-    BIGINT i1 = (BIGINT)std::ceil(kx[i] - ns2); // fine grid start index
-    FLT x1    = (FLT)i1 - kx[i];                // x1 in [-w/2,-w/2+1], up to rounding
+    const auto i1 = BIGINT(std::ceil(kx[i] - ns2)); // fine grid start index
+    // FLT(i1) has different semantics and results an extra cast
+    auto x1 = std::ceil(kx[i] - ns2) - kx[i]; // x1 in [-w/2,-w/2+1], up to rounding
     // However if N1*epsmach>O(1) then can cause O(1) errors in x1, hence ppoly
     // kernel evaluation will fall outside their designed domains, >>1 errors.
     // This can only happen if the overall error would be O(1) anyway. Clip x1??
     if (x1 < -ns2) x1 = -ns2;
     if (x1 > -ns2 + 1) x1 = -ns2 + 1; // ***
-    if (opts.kerevalmeth == 0) {      // faster Horner poly method
-      set_kernel_args(kernel_args, x1, opts);
-      evaluate_kernel_vector(ker, kernel_args, opts, ns);
-    } else
-      eval_kernel_vec_Horner(ker, x1, ns, opts);
-    BIGINT j = i1 - off1; // offset rel to subgrid, starts the output indices
-    // critical inner loop:
-    for (int dx = 0; dx < ns; ++dx) {
-      FLT k = ker[dx];
-      du[2 * j] += re0 * k;
-      du[2 * j + 1] += im0 * k;
-      ++j;
+    //alignas(alignment) const auto ker = ker_eval<ns, kerevalmeth, FLT, batch_t>(opts, x1);
+    const auto &ker = ker_eval<ns, kerevalmeth, FLT, batch_t>(opts, x1);
+    const auto j = i1 - off1;         // offset rel to subgrid, starts the output indices
+    auto *FINUFFT_RESTRICT trg = du + 2 * j; // restrict helps compiler to vectorize
+    // du is padded, so we can use SIMD even if we write more than ns values in du
+    // ker0 is also padded.
+    // regular_part is the largest multiple of 2*ns minus the remainder modulo
```

Comment:
I can't understand this - can you rewrite this as a definition of "regular_part"? (what you wrote states "regular_part" is a multiple of 2ns. Do you mean it is a multiple of avxsize?) Please cite VCL docs and try to rewrite - I know it's hard to explain.

Reply:
Eg after discussion:  padding>=0 chosen so (2ns+padding) = m * avxsize.  Call this L.  If m even, regular_part = L (and chunks of 2avxsize are used).  Else regular_part = L-axvsize, and the last avxsize elements of L are done as special case at the end.